server/cluster.go

```diff
@@ -0,0 +1,288 @@
+		if ok {
+			return errors.Errorf("duplicated store id in %v for bootstrap %d", storesMeta, clusterID)
+		}
+		storeIDs[storeID] = struct{}{}
+
+		if storeMeta.GetNodeId() != nodeMeta.GetNodeId() {
+			return errors.Errorf("invalid store node id %d != %d for bootstrap %d", storeMeta.GetNodeId(), nodeMeta.GetNodeId(), clusterID)
+		}
+
+		if storeID == peer.GetStoreId() {
+			found = true
+		}
+	}
+
+	if !found {
+		return errors.Errorf("invalid peer store id %d, not in %v for bootstrap %d", peer.GetStoreId(), storesMeta, clusterID)
+	}
+
+	return nil
+}
+
+func (s *Server) bootstrapCluster(clusterID uint64, req *protopb.BootstrapRequest) error {
+	log.Infof("try to bootstrap cluster %d with %v", clusterID, req)
+
+	if err := checkBootstrapRequest(clusterID, req); err != nil {
+		return errors.Trace(err)
+	}
+
+	var ops []clientv3.Op
+
+	clusterMeta := protopb.Cluster{
+		ClusterId:     proto.Uint64(clusterID),
+		MaxPeerNumber: proto.Uint32(defaultMaxPeerNumber),
+	}
+
+	// Set cluster meta
+	clusterValue, err := proto.Marshal(&clusterMeta)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	clusterRootPath := s.getClusterRootPath(clusterID)
+	ops = append(ops, clientv3.OpPut(clusterRootPath, string(clusterValue)))
+
+	// Set node meta
+	nodePath := getClusterNodeKey(clusterRootPath, req.GetNode().GetNodeId())
+	nodeValue, err := proto.Marshal(req.GetNode())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ops = append(ops, clientv3.OpPut(nodePath, string(nodeValue)))
+
+	// Set store meta
+	for _, storeMeta := range req.GetStores() {
+		storePath := getClusterStoreKey(clusterRootPath, storeMeta.GetStoreId())
+		storeValue, err1 := proto.Marshal(storeMeta)
+		if err1 != nil {
+			return errors.Trace(err)
+		}
+		ops = append(ops, clientv3.OpPut(storePath, string(storeValue)))
+	}
+
+	// Set region id -> search key
+	regionPath := getClusterRegionKey(clusterRootPath, req.GetRegion().GetRegionId())
+	ops = append(ops, clientv3.OpPut(regionPath, encodeRegionSearchKey(req.GetRegion().GetEndKey())))
+
+	// Set region meta with search key
+	regionSearchPath := getClusterRegionSearchKey(clusterRootPath, req.GetRegion().GetEndKey())
+	regionValue, err := proto.Marshal(req.GetRegion())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ops = append(ops, clientv3.OpPut(regionSearchPath, string(regionValue)))
+
+	bootstrapCmp := clientv3.Compare(clientv3.CreatedRevision(clusterRootPath), "=", 0)
+	resp, err := s.client.Txn(context.TODO()).
+		If(s.leaderCmp(), bootstrapCmp).
+		Then(ops...).
+		Commit()
+	if err != nil {
+		return errors.Trace(err)
+	} else if !resp.Succeeded {
+		return errors.Errorf("bootstrap cluster %d fail, we may not leader", clusterID)
+	}
+
+	log.Infof("bootstrap cluster %d ok", clusterID)
+
+	s.clusterLock.Lock()
+	defer s.clusterLock.Unlock()
+
+	if _, ok := s.clusters[clusterID]; ok {
+		// We have bootstrapped cluster ok, and another goroutine quickly requests use this
+		// so we create the cluster object.
```

Comment:
no, the etcd transaction successes already. 

When we bootstrap in etcd, another thread can use it quickly and if it finds no cluster object is created, it creates. 

But It is nearly impossible to enter this case, I will log it later.